common/lib/xmodule/xmodule/modulestore/split_mongo/split.py

```diff
@@ -1195,31 +1201,51 @@ def _block_matches_all(block_data):
         # don't expect caller to know that children are in fields
         if 'children' in qualifiers:
             settings['children'] = qualifiers.pop('children')
+
+        path_cache = {}
         for block_id, value in course.structure['blocks'].iteritems():
             if _block_matches_all(value):
-                items.append(block_id)
+                if not include_orphans:
+                    if block_id.type in DETACHED_XBLOCK_TYPES or self.has_path_to_root(block_id, course, path_cache):
+                        items.append(block_id)
+                else:
+                    items.append(block_id)
 
         if len(items) > 0:
             return self._load_items(course, items, depth=0, **kwargs)
         else:
             return []
 
-    def has_path_to_root(self, block_key, course):
+    def has_path_to_root(self, block_key, course, path_cache=None):
         """
         Check recursively if an xblock has a path to the course root
 
         :param block_key: BlockKey of the component whose path is to be checked
         :param course: actual db json of course from structures
+        :param path_cache: a dictionary that records which modules have a path to the root so that we don't have to
+        double count modules if we're computing this for a list of modules in a course.
 
         :return Bool: whether or not component has path to the root
         """
 
+        if path_cache and block_key in path_cache:
+            return path_cache[block_key]
+
```

Comment:
@ormsbee I think, I am missing what you are trying to convey here. `path_cache` is being initialized in `get_items` if `include_orphans` is `False`. Also, why do we need to initialize `path_cache` if its None (as this is not going to be used).